docs/api-reference/core/resource-management/program-manager.md

```diff
@@ -0,0 +1,104 @@
+# ProgramManager
+
+The `ProgramManger` manages the creation and caching of programs. It allows the application to request a program based on a vertex shader, fragment shader and set of defines, modules and code injections. The `ProgramManager` will return the requested program, creating it the first time, and re-using a cached version if it is requested more than once. It also allows for the definition of hook functions and module code injections to be inserted into shaders.
+
+
+## Usage
+
+```js
+const pm = new ProgramManager(gl);
+
+const vs = `
+attribute vec4 position;
+
+void main() {
+#ifdef MY_DEFINE
+  gl_Position = position;
+#else
+  gl_Position = position.wzyx;
+#endif
+}
+`;
+
+const fs = `
+void main() {
+  gl_FragColor = vec4(1.0, 0.0, 0.0, 1.0);
+  MY_SHADER_HOOK(gl_FragColor);
+}
+`;
+
+pm.addShaderHook('fs:MY_SHADER_HOOK(inout vec4 color)');
+
+pm.addModuleInjection(picking, {
+  hook: 'fs:#MY_SHADER_HOOK',
+  injection: 'color = picking_filterColor(color);',
+  order: Number.POSITIVE_INFINITY
+});
+
+const program1 = pm.get(vs, fs);   // Basic, no modules or defines
+const program2 = pm.get(vs, fs);   // Cached, same as program 1, use count 2
+const program3 = pm.get(vs, fs, {  // New program, with different source based on define
```

Comment:
Can we align this API with that of Model, Program and Shader (fs and vs as part of the named parameters).

Reply:
Sure, that makes sense.